### PR TITLE
urls: use https, follow perm redir, fix broken links (in Formulae)

### PR DESCRIPTION
**Pull request by vszakats**
most hand verified using browser (FF 34) and curl

### CODE REVIEW

**Comment by mikemcquaid** [comment]
Needs rebased on `master`. Also, which ones weren't verified?

File: Library/Formula/ppss.rb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 require 'formula'
 
 class Ppss < Formula
-  homepage 'http://ppss.googlecode.com'
+  homepage 'http://code.google.com/p/ppss/'
```
Comment (bfontaine) [inline Library/Formula/ppss.rb]: Shouldn’t it be "https://code.google.com/p/ppss/"?

**Comment by vszakats** [comment]
Not verified initially: those which followed a very clear scheme, for example I didn't verify each language of the `aspell` packages. Verified all homepages, comment URLs (except obvious ones like Google Groups) and repo heads. Now I went on to verify all `url` and `mirror` links using a script. Found one, which was broken before this patch. The rest was fine.

Working on the rebase.

File: Library/Formula/ppss.rb (hunk 1 above)
Comment (vszakats) [inline Library/Formula/ppss.rb]: Indeed, thanks. It's a page that has mixed content, so tried it with plain http to no avail, missed to put back https.

I didn't update Google Code links to https that already use above new naming scheme, _and_ have mixed-content - even though modern browsers will redirect to https automatically anyway. Maybe it should be done.

**Comment by vszakats** [comment]
Rebased, two problems (old broken url and missing https in ppss) fixed, squashed.

**Comment by DomT4** [comment]
My only concern with doing this in such a large batch is that users are going to see seemingly 280 formula with updates pending and freak.

**Comment by vszakats** [comment]
I have another fresh batch of about 150 formula (all hand checked). I can push that here, and if you find it okay, I squash and the patch then can be split up to as many chunks as you guys prefer. Next concern is that they will quickly start to rot as formulas get their normal updates.

**Comment by DomT4** [comment]
My concern is that a lot of people don't update super-regularly, so when they run ` brew update ` they could see pages and pages of updates, and that tends to scare people off ` brew upgrade ` and when that happens, things break.

IMO, I'd love to start committing largely stylistic changes to formulae, to reduce the strict audit fails, to improve security on links that don't automatically redirect to SSL/TLS, but before we do so I'd like to see some mechanism around maintainers being able to exclude or include changes on the ` brew update ` summary, something like so we can add the words ` IncludeInChanges ` or ` ExcludeInChanges ` into the commit message and that would dictate to ` brew update ` whether to display the change or not.

**Comment by bfontaine** [comment]
@DomT4 on this subject, see #35468 (which itself references #25951 in the comments)

**Comment by vszakats** [comment]
I'd argue these changes are not merely stylistic (above the few actual broken links that got fixed). SSL has multiple advantages even if most of the downloaded code is explicitly verified against their checksums (but volatile - `head` - entries and externally stored patches are not). As a Homebrew user it's also concerning to watch how many package still uses plain HTTP, plain GIT and even FTP to pull code to be executed on a trusted local machine.

Anyhow, I'm adding the second patch and you decide.

**Comment by DomT4** [comment]
FWIW, I don't think you'll find anyone at Homebrew who is more paranoid about security and jumpy about unencrypted connections than I am :grinning:, I'm pretty sure the maintainers consider my obsessive security tweaks with more regard for security than end-user convenience pretty tedious most of the time, heh.

As something of a compromise, I'd propose the following:

* Our new Rubocop should enforce SSL/TLS links on the links we absolutely know don't redirect to SSL/TLS. (Such as GNU). I don't think that'd be overly difficult to write into the Rubocop, if the maintainers don't hate the idea.

* Things that automatically redirect from 80 to 443 are largely stylistic, not security, and should be dealt with outside of the security bracket. This includes Wordpress, Debian, code.google, etc which you can't access unencrypted. Sadly, I appreciate not many people have that permanent redirect in place yet.

* Changing the Sourceforge homepages to ` https ` does nothing, because SF have a hard redirect from SSL/TLS to HTTP for non-download links. We also need to be careful with the ` github.io ` links because people are *terrible* at serving mixed content, which breaks SSL/TLS, and often, breaks the webpage to boot.

I'm not against this idea at all, I think the changes are by and large great, and well overdue, but I am incredibly worried about people doing ` brew update `, seeing 450+ updates and going ` hell no `.

**Comment by vszakats** [comment]
About sourceforge - I know :( my only (tapering) hope is that eventually they will make the switch to https once enough clients attempts to use it. I opted to add it because existing state was a seemingly random mixture of http/https, and I think uniform https is better than that. Now that it's uniform, a few s&r can switch it back to plaintext.

**Comment by vszakats** [comment]
The `github.io` mixed content issue effected those sites which were patched up in the second round, of these there is only one where it resulted in a visibly broken page, I left that (`cereal`) plaintext.

**Comment by vszakats** [comment]
"Things that automatically redirect from 80 to 443 are largely stylistic": The traffic on port 80 is leaking and can also be manipulated, to the extent of altering or even wholly preventing the redirect, so IMO this is unsafe. To keep a compromise in context of this patch, those that involve a human could be reverted to http. These are `homepage` and comments/readable output.

**Comment by vszakats** [comment]
After deleting hunks that involved human-readable, forced redirects or mere protocol change for sourceforge addresses, here is the resulting new branch/patch:

https://github.com/vszakats/homebrew/commit/c1e6b147bb4f690f232830e12921831e7c574368

Total number of changed formula went down to 269 from 423.

**Comment by DomT4** [comment]
I've pushed what can either be a companion or an alternative to this in #35551, depending on how the maintainers feel. I'd like to get #35551 committed regardless of what happens to this PR, because it can be complimentary to this, and it prevents this thing creeping back in over time.

@jacknagel, @mikemcquaid - Whichever one of you sees this first, the bot will need killing here again or it'll have a meltdown :grinning:.

**Comment by vszakats** [comment]
Shoved off 3 more files:
https://github.com/vszakats/homebrew/commit/ddd88c334c1ab04d7df2dd1a86c086b198d66dda

**Comment by DomT4** [comment]
> About sourceforge - I know :( my only (tapering) hope is that eventually they will make the switch to https once enough clients attempt to use it (it is in fact used already if you're logged in to their service though).

Can't even begin to say how much it annoys me that they offer SSL/TLS as a user privilege and not a general thing.

It's probably worth waiting for a maintainer to drop by before pushing additional changes here so they can kill the bot checks before they begin; the way the bot works is it'll grab any PR or new commit submitted, and start going through it for errors. Obviously, having to build and test some 200+ formulae is going to give the poor bot something of a personal crisis, heh :wink:.

**Comment by vszakats** [comment]
Oh, I was wondering what causes the large bot-headache. Someone better be pausing it, or, if there exists a silencer keyword to put inside the commit msg, I'd happily start using it. (didn't find such in `/share/doc/Brew-Test-Bot*.md`)

sf.net: I'm guessing there is some non-trivial stuff to solve around their hand-rolled mirroring infrastructure. Similarly for the user pages hosted as subdomains. That still doesn't explain the lacking ssl/tls on naked 'sourceforge.net' urls though.

**Comment by DomT4** [comment]
Yeah, there's no existing keyword yet for ` Don't check this ` or ` Check this `. It just checks everything, which works most of the time, but sometimes large PRs creep up and the bot just baaaaaaacks up for hours and hours on end, heh. They're fast machines, in a fast data center, but still, lot of changes to process and test. Mike seems to be improving the mechanism constantly though, so it's not a finished product in terms of x will always be x in x situation.

> Similarly for the user pages hosted as subdomains.

This I can understand. Frankly, given the state of most of the user-created subdomain pages, slapping SSL/TLS on those without warning could actually break people's websites horribly. It's easier to write SSL/TLS into a website from the beginning than having to change infrastructure to do so later, alas. But yes, if they can do full-site, strong SSL for what we use as homepage and bug report links as long as the user is logged in, there's little excuse for not rolling that out to everyone :disappointed:.

**Comment by vszakats** [comment]
Not only the user's own website, but the host website as well, through f.e. CSRF attacks. Here's how Github solved it some time ago:
https://github.com/blog/1452-new-github-pages-domain-github-io

[ Opened a ticket: https://sourceforge.net/p/forge/site-support/9353/ ]

**Comment by jacknagel** [comment]
Let's avoid pushing to this PR any further, it's going to wreak havoc on CI.

Maybe just start a new branch and link to it.

**Comment by vszakats** [comment]
@jacknagel It's already done that way, see above: https://github.com/Homebrew/homebrew/pull/35545#issuecomment-68655475

Linke to the branch: https://github.com/vszakats/homebrew/commit/ddd88c334c1ab04d7df2dd1a86c086b198d66dda

**Comment by vszakats** [comment]
Here's the list of HTTPS preload list for Chrome/Firefox browsers. It can be used to expand the list of URLs where HTTPS is enforced in formulae, too:
https://chromium.googlesource.com/chromium/src/+/master/net/http/transport_security_state_static.json

Here's the database of HTTPS Everywhere by EFF:
https://github.com/EFForg/https-everywhere/tree/master/src/chrome/content/rules

Search interface for the above:
https://www.eff.org/https-everywhere/atlas/

Having said these, I'm closing this pull.

**Pull request closed without merging**